File: proxy/myproxy.cpp
#include <iostream>

using namespace std;

class Burger {
    public:
    virtual void Prepare() const = 0;
};

class VegBurger : public Burger {
    public:
    void Prepare() const override {
        cout << "Preparing Veggie Burger" << endl;
    }
};

class BurgerProxy : public Burger {
    private:
    VegBurger *vegburger;

    bool TastesGood() const {
        // real: interact with veg burger
        return true;
    }

    bool IsReady() const {
        // real: interact with veg burger
        return true;
    }

    public:
    BurgerProxy(VegBurger *vegburger) : vegburger(new VegBurger(*vegburger)) {
    }

    ~BurgerProxy() {
        delete vegburger;
    }

    void Prepare() const override {
        this->vegburger->Prepare();
        if (this->IsReady()) {
            if (this->TastesGood()) {
                cout << "Ready and Tasty !" << endl;
            }
        }
    }
};

void client(const Burger &burger) {
    burger.Prepare();
}

int main() {
    VegBurger *vegburger = new VegBurger();
    client(*vegburger);

    BurgerProxy *burgerproxy = new BurgerProxy(vegburger);
    client(*burgerproxy);

    delete vegburger;
    delete burgerproxy;

    return(0);
}

File: proxy/swift/myproxy.swift
protocol Burger {
    func Prepare() -> String
}

struct VegBurger : Burger {
    func Prepare() -> String {
        return("VegBurger : prepare")
    }
}

struct BurgerProxy : Burger {
    var burger : Burger

    init(burger : Burger) {
        self.burger = burger
    }

    func Prepare() -> String {
        if TastesGood() && IsHealthy() {
            print(burger.Prepare())
            return "BurgerProxy : prepare"
        } else {
            return "BurgerProxy : Bad Burger"
        }
    }

    func TastesGood() -> Bool {
        return true
    }

    func IsHealthy() -> Bool {
        return true
    }
}

@main
class MyProxy {
    static func main() {
        let vegburger = VegBurger()
        print(vegburger.Prepare())

        let someburger = BurgerProxy(burger: vegburger)
        print(someburger.Prepare())
    }
}